lib/_http_outgoing.js

```diff
@@ -70,6 +70,27 @@ function OutgoingMessage() {
   this._header = null;
   this._headers = null;
   this._headerNames = {};
+
+  var self = this;
+
+  // Call remaining callbacks if stream closes prematurely
+  this.once('close', function() {
+    if (self.output.length === 0) return;
+
+    var err = new Error('stream closed prematurely');
+
+    var outputCallbacks = self.outputCallbacks;
+
+    self.output = [];
+    self.outputEncodings = [];
+    self.outputCallbacks = [];
+
+    for (var i = 0; i < outputCallbacks.length; i++) {
```

Comment:
I worry the callback function over which we have no control might do something fishy and mess up our state while we're working with it. If such things are nothing we are supposed to be guarding against, then I'll change it of course.